[OpenMP] OpenMP 5.1 "assume" directive parsing support

jtb20 opened the pull request:
This is a minimal patch to support parsing for "omp assume" directives. These are meant to be hints to a compiler's optimisers: as such, it is legitimate (if not very useful) to ignore them.  The patch builds on top of the existing support for "omp assumes" directives (note spelling!).

Unlike the "omp [begin/end] assumes" directives, "omp assume" is associated with a compound statement, i.e. it can appear within a function.  The "holds" assumption could (theoretically) be mapped onto the existing builtin "__builtin_assume", though the latter applies to a single point in the program, and the former to a range (i.e. the whole of the associated compound statement).

This patch fixes sollve's OpenMP 5.1 "omp assume"-based tests.

jtb20 commented:
Adding @jdoerfert as potential reviewer

jtb20 commented:
Ping?

shiltian commented on clang/test/OpenMP/assume_lambda.cpp:
```diff
@@ -0,0 +1,31 @@
+// CHECK: auto fn = [](int x) {
+    return fn(5);
+  }
+}
```
empty line EoF

jtb20 replied:
Ugh, a bad habit of that editor. I guess I should try to figure out a way of stopping it doing that automatically...

shiltian replied:
I mean, we need an empty line at the end of file.

jtb20 replied:
Understood! There is indeed a vscode option to add the missing newline (it appears to be turned off by default for some bizarre reason). I'll push a new version with them in.

shiltian commented on clang/test/OpenMP/assume_template.cpp:
```diff
@@ -0,0 +1,42 @@
+  }
+}
+
+#endif
```
ditto

shiltian reviewed (commented):
don't you need more code in AST?

jtb20 commented:
> don't you need more code in AST?

Sorry, I don't quite understand the question! Could you elaborate a little please?

shiltian commented:
> > don't you need more code in AST?
> 
> Sorry, I don't quite understand the question! Could you elaborate a little please?

I was thinking maybe you need changes in AST related files, like `ASTWriter.cpp`, but that might be not needed as this is adding a new directive.

jtb20 commented:
> > > don't you need more code in AST?
> > 
> > 
> > Sorry, I don't quite understand the question! Could you elaborate a little please?
> 
> I was thinking maybe you need changes in AST related files, like `ASTWriter.cpp`, but that might be not needed as this is adding a new directive.

At the moment, since the "assume" directive is parsed but then immediately discarded, I don't think anything else is needed.

Actually the existing "assumes" support is reused -- the bit in SemaOpenMP.cpp adds the "assume" assumptions to the OMPAssumeScoped "stack". For "assumes", it's done like that so (e.g. top-level) declarations between begin/end "assumes" can be modified according to the assumptions on that stack. For "assume", once some use is made for the assumptions, that might turn out to not be the most useful representation. That can be revisited later though, I think.

alexey-bataev commented:
> > > > don't you need more code in AST?
> > > 
> > > 
> > > Sorry, I don't quite understand the question! Could you elaborate a little please?
> > 
> > 
> > I was thinking maybe you need changes in AST related files, like `ASTWriter.cpp`, but that might be not needed as this is adding a new directive.
> 
> At the moment, since the "assume" directive is parsed but then immediately discarded, I don't think anything else is needed.
> 
> Actually the existing "assumes" support is reused -- the bit in SemaOpenMP.cpp adds the "assume" assumptions to the OMPAssumeScoped "stack". For "assumes", it's done like that so (e.g. top-level) declarations between begin/end "assumes" can be modified according to the assumptions on that stack. For "assume", once some use is made for the assumptions, that might turn out to not be the most useful representation. That can be revisited later though, I think.



> > > > don't you need more code in AST?
> > > 
> > > 
> > > Sorry, I don't quite understand the question! Could you elaborate a little please?
> > 
> > 
> > I was thinking maybe you need changes in AST related files, like `ASTWriter.cpp`, but that might be not needed as this is adding a new directive.
> 
> At the moment, since the "assume" directive is parsed but then immediately discarded, I don't think anything else is needed.
> 
> Actually the existing "assumes" support is reused -- the bit in SemaOpenMP.cpp adds the "assume" assumptions to the OMPAssumeScoped "stack". For "assumes", it's done like that so (e.g. top-level) declarations between begin/end "assumes" can be modified according to the assumptions on that stack. For "assume", once some use is made for the assumptions, that might turn out to not be the most useful representation. That can be revisited later though, I think.

Even so, it should support serialization/deserialization and have the tests for it (like ast-dump and ast-print).

alexey-bataev commented on clang/include/clang/AST/OpenMPClause.h:
```diff
@@ -2013,6 +2014,179 @@ class OMPMergeableClause : public OMPClause {
+/// In this example directive '#pragma omp assume' has a 'contains' clause.
+class OMPContainsClause final
+    : public OMPNoChildClause<llvm::omp::OMPC_contains> {
+  llvm::SmallSet<OpenMPDirectiveKind, 4> DirectiveKinds;
```
Bad approach, use tail allocation to keep this

jtb20 replied:
To clarify: do you mean that I should use "ompchildren" and encode the directive kind list as OMPClauses, rather than use a separate allocation/data structure?

That would probably work I guess, but it means adding a pseudo-clause type (OMPC_directive_kind or similar), or adding all the possible directive names as clause names also (which seems unwieldy for a relatively obscure feature). Is one of those along the lines of what you were thinking?

alexey-bataev replied:
I mean that dynamic data types (do vectors, sets, etc.) are not allowed in AST

alexey-bataev commented on clang/include/clang/AST/OpenMPClause.h:
```diff
@@ -2013,6 +2014,179 @@ class OMPMergeableClause : public OMPClause {
+/// \endcode
+/// In this example directive '#pragma omp assume' has an 'absent' clause.
+class OMPAbsentClause final : public OMPNoChildClause<llvm::omp::OMPC_absent> {
+  llvm::SmallSet<OpenMPDirectiveKind, 4> DirectiveKinds;
```
Must be tail allocated

alexey-bataev commented on clang/include/clang/AST/OpenMPClause.h:
```diff
@@ -342,6 +342,57 @@ template <class T> class OMPVarListClause : public OMPClause {
   }
 };
 
+template <class T> class OMPDirectiveListClause : public OMPClause {
```
Why do you need this?

jtb20 replied:
The idea was to factor out common functionality shared by absent/contains clauses. It doesn't do a huge amount, admittedly, but I think it's still better than copy/pasting a bunch of code for the absent/contains clause classes. (OMPNoChildClause, OMPOneStmtClause are similar in intent though perhaps wider in applicability).

It should probably have a comment, if that's what you meant!

alexey-bataev commented on clang/include/clang/AST/OpenMPClause.h:
```diff
@@ -2013,6 +2064,184 @@ class OMPMergeableClause : public OMPClause {
+  friend TrailingObjects;
+
+public:
+  /// Build 'absent' clause.
+  ///
+  /// \param StartLoc Starting location of the clause.
+  /// \param LParenLoc Location of '('.
+  /// \param EndLoc Ending location of the clause.
+  /// \param NumKinds Number of directive kinds listed in the clause.
+  OMPAbsentClause(SourceLocation StartLoc, SourceLocation LParenLoc,
+                  SourceLocation EndLoc, unsigned NumKinds)
+      : OMPDirectiveListClause<OMPAbsentClause>(
+            llvm::omp::OMPC_absent, StartLoc, LParenLoc, EndLoc, NumKinds) {}
+
+  /// Build an empty clause.
+  OMPAbsentClause(unsigned NumKinds)
+      : OMPDirectiveListClause<OMPAbsentClause>(
+            llvm::omp::OMPC_absent, SourceLocation(), SourceLocation(),
+            SourceLocation(), NumKinds) {}
+
```
These constructions must be private/protected

alexey-bataev commented on clang/include/clang/AST/OpenMPClause.h:
```diff
@@ -2013,6 +2064,184 @@ class OMPMergeableClause : public OMPClause {
+  friend TrailingObjects;
+
+public:
+  /// Build 'contains' clause.
+  ///
+  /// \param StartLoc Starting location of the clause.
+  /// \param LParenLoc Location of '('.
+  /// \param EndLoc Ending location of the clause.
+  /// \param NumKinds Number of directive kinds listed in the clause.
+  OMPContainsClause(SourceLocation StartLoc, SourceLocation LParenLoc,
+                    SourceLocation EndLoc, unsigned NumKinds)
+      : OMPDirectiveListClause<OMPContainsClause>(
+            llvm::omp::OMPC_contains, StartLoc, LParenLoc, EndLoc, NumKinds) {}
+
+  /// Build an empty clause.
+  OMPContainsClause(unsigned NumKinds)
+      : OMPDirectiveListClause<OMPContainsClause>(
+            llvm::omp::OMPC_contains, SourceLocation(), SourceLocation(),
+            SourceLocation(), NumKinds) {}
+
```
Private or protected constructors

alexey-bataev commented on clang/include/clang/AST/OpenMPClause.h:
```diff
@@ -2013,6 +2064,184 @@ class OMPMergeableClause : public OMPClause {
+/// \endcode
+/// In this example directive '#pragma omp assume' has a 'no_openmp_routines'
+/// clause.
+class OMPNoOpenMPRoutinesClause final
```
Split the patch into many patches, one for each particular clause.

jtb20 replied:
I understand splitting patches into logical units, but that seems a bit unnecessary. What's the motivation? The mid-way point (half the clauses supported, say) isn't a useful state.

alexey-bataev replied:
https://llvm.org/docs/DeveloperPolicy.html#id22

jtb20 committed:
[OpenMP][clang] New tests for 'omp assume' directive

jtb20 committed:
[OpenMP][clang] Add 'holds' clause for 'omp assume' directive

jtb20 committed:
[OpenMP][clang] Add 'absent' clause for 'omp assume' directive

jtb20 committed:
[OpenMP][clang] Add 'contains' clause for 'omp assume' directive

jtb20 committed:
[OpenMP][clang] Add 'no_openmp' clause for 'omp assume' directive

jtb20 committed:
[OpenMP][clang] Add 'no_openmp_routines' clause for 'omp assume' directive

jtb20 committed:
[OpenMP][clang] Add 'no_parallelism' clause for 'omp assume' directive

jtb20 committed:
[OpenMP][clang] OpenMP 5.1 "assume" directive parsing support

This patch supports parsing of "omp assume" directives.  These are meant
to be hints to a compiler's optimisers: as such, it is legitimate
(if not very useful) to ignore them.  This version of the patch
implements new AST nodes for the "assume" directive and the various
clauses it accepts as arguments, and adds new (C++ module) tests for
serialization/deserialization of said nodes.

The patch now uses tail allocation for the directive kind list.

Unlike the "omp [begin/end] assumes" directives, "omp assume" is
associated with a compound statement, i.e. it can appear within a
function.  The "holds" assumption could (theoretically) be mapped
onto the existing builtin "__builtin_assume", though the latter applies
to a single point in the program, and the former to a range (i.e. the
whole of the associated compound statement).

The "assume" directive appears to be distinct from the [[omp::assume]]
annotation.

This patch fixes sollve's OpenMP 5.1 "omp assume"-based tests.

jtb20 commented:
The compiler builds for each intermediate step, but the result isn't necessarily useful until the 'omp assume' patch.

alexey-bataev reviewed (commented):
Need a test with the nesting of the directives

alexey-bataev commented on clang/include/clang/Parse/Parser.h:
```diff
@@ -3532,6 +3532,10 @@ class Parser : public CodeCompletionHandler {
                                  OpenMPDirectiveKind DKind, SourceLocation Loc,
                                  bool ReadDirectiveWithinMetadirective);
 
+  StmtResult ParseOpenMPInformationalDirective(
```
Add a comment for the function

alexey-bataev commented on clang/include/clang/Sema/SemaOpenMP.h:
```diff
@@ -399,6 +399,14 @@ class SemaOpenMP : public SemaBase {
       OpenMPDirectiveKind Kind, const DeclarationNameInfo &DirName,
       OpenMPDirectiveKind CancelRegion, ArrayRef<OMPClause *> Clauses,
       Stmt *AStmt, SourceLocation StartLoc, SourceLocation EndLoc);
+  StmtResult ActOnOpenMPInformationalDirective(
```
Add a comment for the function

alexey-bataev commented on clang/include/clang/Sema/SemaOpenMP.h:
```diff
@@ -399,6 +399,14 @@ class SemaOpenMP : public SemaBase {
+      OpenMPDirectiveKind Kind, const DeclarationNameInfo &DirName,
+      ArrayRef<OMPClause *> Clauses, Stmt *AStmt, SourceLocation StartLoc,
+      SourceLocation EndLoc);
+  StmtResult ActOnOpenMPAssumeDirective(ArrayRef<OMPClause *> Clauses,
```
Add a comment for the function

alexey-bataev commented on clang/lib/Sema/TreeTransform.h:
```diff
@@ -9193,6 +9207,63 @@ StmtResult TreeTransform<Derived>::TransformOMPExecutableDirective(
+  llvm::SmallVector<OMPClause *, 16> TClauses;
+  ArrayRef<OMPClause *> Clauses = D->clauses();
+  TClauses.reserve(Clauses.size());
+  for (ArrayRef<OMPClause *>::iterator I = Clauses.begin(), E = Clauses.end();
+       I != E; ++I) {
```
```suggestion
  for (OMPClause *C : Clauses) {
```

alexey-bataev commented on clang/lib/Sema/TreeTransform.h:
```diff
@@ -9193,6 +9207,63 @@ StmtResult TreeTransform<Derived>::TransformOMPExecutableDirective(
+    {
+      Sema::CompoundScopeRAII CompoundScope(getSema());
+      Stmt *CS = nullptr;
+      if (D->getDirectiveKind() == OMPD_assume)
+        CS = D->getAssociatedStmt();
+      else
+        llvm_unreachable("Unexpected informational directive");
```
```suggestion
      assert(D->getDirectiveKind() == OMPD_assume && "Unexpected informational directive");  
      CS = D->getAssociatedStmt();
```

alexey-bataev commented on clang/lib/Sema/TreeTransform.h:
```diff
@@ -9193,6 +9207,63 @@ StmtResult TreeTransform<Derived>::TransformOMPExecutableDirective(
+    }
+    AssociatedStmt =
+        getDerived().getSema().OpenMP().ActOnOpenMPRegionEnd(Body, TClauses);
+    if (AssociatedStmt.isInvalid()) {
+      return StmtError();
+    }
```
Drop extra braces

alexey-bataev commented on clang/lib/Sema/TreeTransform.h:
```diff
@@ -9193,6 +9207,63 @@ StmtResult TreeTransform<Derived>::TransformOMPExecutableDirective(
+      return StmtError();
+    }
+  }
+  if (TClauses.size() != Clauses.size()) {
+    return StmtError();
+  }
```
Drop extra braces

alexey-bataev reviewed (commented):
Add the notest to OpenMPSupport.rst and release notes

alexey-bataev reviewed (approved):
LG

jtb20 commented:
Thank you for the review! Again, I do not have commit access, so please apply for me if possible.

Julian

jdoerfert commented:
This broke the Flang build. You need to add some Flang code for all changes in omp.td. Look for a similar clause in the flang subproject to see what to do, and verify by building Flang. Maybe revert for now until fixed.

jtb20 commented:
Hopefully fixed by: https://github.com/llvm/llvm-project/pull/102008

clementval commented:
This is still breaking a buildbot. https://lab.llvm.org/buildbot/#/builders/157/builds/4246

Are you working on a fix?

jtb20 commented:
On 06/08/2024 19:40, Valentin Clement (バレンタイン クレメン) wrote:
> 	
> This is still breaking a buildbot.
> https://lab.llvm.org/buildbot/#/builders/157/builds/4246
> <https://lab.llvm.org/buildbot/#/builders/157/builds/4246>
> 
> Are you working on a fix?

It seems to be the same bug -- I assumed the buildbot picked up a
pre-fixed revision.  Is that not the case?

Thanks,

Julian

clementval commented:
Yeah it was probably the case. It's green now.

jtb20 commented:
Thank you for the update!

TigerTimeJ commented on llvm/include/llvm/Frontend/OpenMP/OMP.td:
```diff
@@ -510,6 +528,18 @@ def OMP_EndAssumes : Directive<"end assumes"> {
   let category = OMP_Assumes.category;
 }
+def OMP_Assume : Directive<"assume"> {
+  let association = AS_Block;
```
why need this?

jtb20 replied:
I don't understand the question. Do you think the .td fragment is unnecessary for adding support for the "omp assume" directive? There is no comment here, but that follows the general trend in that file.
